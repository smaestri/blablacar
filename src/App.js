import React, { useEffect, useState } from 'react';
import axios from 'axios';
import ReactPaginate from 'react-paginate';

import 'bootstrap/dist/css/bootstrap.min.css';


const URL = "https://public-api.blablacar.com/api/v3/trips?key=YzbiA8L6DcqxTvSna1lOFQQU66FosDVs&from_coordinate=48.8566%2C2.3522&to_coordinate=45.764043%2C4.835659&from_country=FR&to_country=FR&locale=fr-FR&start_date_local=2021-10-10T00:00:00&currency=EUR&count=100";

function App() {


  const [trips, setTrips] = useState([])
  const [currentTrips, setCurrentTrips] = useState([])
  const [pageCount, setPagecount] = useState(0)

  useEffect(()=> {
    console.log('useeffect')
    axios.get(URL).then(response => {
      console.log('response' + JSON.stringify(response.data))
      const trips = response.data.trips;
      const tripsWithIndex = trips.map((trip, index) => ({ ...trip, index:  index +1 }));
      setTrips(tripsWithIndex);
      setCurrentTrips(tripsWithIndex.slice(0, 5))
      setPagecount(Math.ceil(trips.length / 5))
      console.log('pageCount' + pageCount)
    })
    .catch((error) => {
      alert('error')
    })

  }, [])


  const handlePageClick = (props) => {
    console.log('selected page' + props.selected)
    setCurrentTrips(trips.slice(props.selected *5, ( 5*props.selected) + 5 ))
  }

console.log('trips: ' + JSON.stringify(trips))
  return (
    <div className="App">
      <table border="1"><tr style={{fontWeight: 'bold'}}><td>Num</td><td>From</td><td>To</td><td>Price</td></tr>
      {currentTrips.map((trip) => (
        <tr>
          <td>{trip.index}</td>
          <td>{trip.waypoints[0].place.city}</td>
          <td>{trip.waypoints[1].place.city}</td>
          <td>{`${trip.price.amount} ${trip.price.currency}` }</td>

          </tr>
        ))}
      </table>
      <div id="react-paginate">
      <ReactPaginate
          pageCount={pageCount}
          marginPagesDisplayed={0}
          pageRangeDisplayed={5}
          onPageChange={handlePageClick}
          breakClassName={'page-item'}
 breakLinkClassName={'page-link'}
 containerClassName={'pagination'}
 pageClassName={'page-item'}
 pageLinkClassName={'page-link'}
 previousClassName={'page-item'}
 previousLinkClassName={'page-link'}
 nextClassName={'page-item'}
 nextLinkClassName={'page-link'}
 activeClassName={'active'}
        />
        </div>
<div>
DEBUG
<table border="1"><tr style={{fontWeight: 'bold'}}><td>Num</td><td>From</td><td>To</td><td>Price</td></tr>
      {trips.map((trip, index) => (
        <tr>
          <td>{index+1}</td>
          <td>{trip.waypoints[0].place.city}</td>
          <td>{trip.waypoints[1].place.city}</td>
          <td>{`${trip.price.amount} ${trip.price.currency}` }</td>

          </tr>
        ))}
      </table>
</div>



    </div>
  );
}

export default App;
