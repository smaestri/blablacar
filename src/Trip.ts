
export interface Trip {
  id: number;
  waypoints: Array<WayPoint>;
  price: Price;
}

interface WayPoint {
    place: Place;
}

interface Price {
    amount: number;
    currency: string;
}

interface Place {
    city: string;
}
