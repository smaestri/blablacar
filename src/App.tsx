import React, { useEffect, useState } from 'react';
import axios from 'axios';
import ReactPaginate from 'react-paginate';
import 'bootstrap/dist/css/bootstrap.min.css';
import { Trip } from './Trip';

const URL : string = "https://public-api.blablacar.com/api/v3/trips?key=YzbiA8L6DcqxTvSna1lOFQQU66FosDVs&from_coordinate=48.8566%2C2.3522&to_coordinate=45.764043%2C4.835659&from_country=FR&to_country=FR&locale=fr-FR&start_date_local=2021-10-10T00:00:00&currency=EUR&count=100";
const ITEMS_PER_PAGE : number = 5;

function App() {

  const [trips, setTrips] = useState<Array<Trip>>([])
  const [currentTrips, setCurrentTrips] = useState<Array<Trip>>([])
  const [pageCount, setPagecount] = useState<number>(0)

  useEffect(() => {
    axios.get(URL).then((response:any) => {
      const trips: Array<Trip> = response.data.trips;
      const tripsWithIndex : Array<Trip> = trips.map((trip, index) => ({ ...trip, id: index + 1 }));
      setTrips(tripsWithIndex);
      setCurrentTrips(tripsWithIndex.slice(0, ITEMS_PER_PAGE))
      setPagecount(Math.ceil(trips.length / ITEMS_PER_PAGE))
    })
      .catch((error) => {
        console.log('An error occured + ')
      })

  }, [])


  const handlePageClick = (props: any) => {
    const beginIndex = props.selected * ITEMS_PER_PAGE;
    setCurrentTrips(trips.slice(beginIndex, beginIndex + ITEMS_PER_PAGE))
  }

  return (
    <div className="container">

      <h2>Trips from Paris to Lyon</h2>

      <table className="table"><thead><tr><th scope="col">Num</th><th scope="col">From</th><th scope="col">To</th><th scope="col">Price</th></tr></thead>
        {currentTrips.map((trip: Trip) => (
          <tbody>
          <tr>
            <th scope="row">{trip.id}</th>
            <td>{trip.waypoints[0].place.city}</td>
            <td>{trip.waypoints[1].place.city}</td>
            <td>{`${trip.price.amount} ${trip.price.currency}`}</td>
          </tr>
          </tbody>
        ))}
      </table>
        <ReactPaginate
          pageCount={pageCount}
          marginPagesDisplayed={1}
          pageRangeDisplayed={ITEMS_PER_PAGE}
          onPageChange={handlePageClick}
          breakClassName={'page-item'}
          breakLinkClassName={'page-link'}
          containerClassName={'pagination justify-content-center'}
          pageClassName={'page-item'}
          pageLinkClassName={'page-link'}
          previousClassName={'page-item'}
          previousLinkClassName={'page-link'}
          nextClassName={'page-item'}
          nextLinkClassName={'page-link'}
          activeClassName={'active'}
        />
    </div>
  );
}

export default App;
