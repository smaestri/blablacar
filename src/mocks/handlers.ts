import { rest } from 'msw'
import { Trip } from '../Trip'

export const handlers = [
    rest.get('https://public-api.blablacar.com/api/v3/trips', (req, res, ctx) => {
      const result : {trips: Array<Trip>} = {
        trips: [{id:1,
          price: {amount: 10, currency: 'EUR'},
          waypoints: [
            {place: {city: 'Paris'}},
            {place: {city: 'Lyon'}}]
       }]
      }
        return res(
            ctx.status(200),
            ctx.json(result)
        )
    })
  ]