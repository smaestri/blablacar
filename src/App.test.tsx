import { render, screen, waitFor } from '@testing-library/react';
import { rest } from 'msw';
import App from './App';
import { server } from './mocks/server';
import { Trip } from './Trip';

const createTrips = (count: number) => {
  const res: Array<Trip> = []
  for (let i = 0;i < count; i++) {
    res.push(
      {
        id: i,
        price: { amount: i * 10, currency: 'EUR' },
        waypoints: [
          { place: { city: 'Paris' + i } },
          { place: { city: 'Lyon' + i } }]
      }
    )
  }
  return res;
}

test('renders without pagination', async () => {
 
  // when
  const {container, getByText} = render(<App />);
  await waitFor(() => getByText("Paris"))

  // then
  expect(container).toMatchSnapshot();
});

test('renders with pagination', async () => {
  // given
  const trips = createTrips(100);
  //override mock
  server.use(
    rest.get('https://public-api.blablacar.com/api/v3/trips', (req, res, ctx) => {
      const result: { trips: Array<Trip> } = {
        trips 
      }
      return res(
        ctx.status(200),
        ctx.json(result)
      )
    })
  )
  // when
  const {container, getByText} = render(<App />);
  await waitFor(() => getByText("Paris1"))

  // then
  expect(container).toMatchSnapshot();
});
