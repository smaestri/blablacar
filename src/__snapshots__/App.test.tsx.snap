// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with pagination 1`] = `
<div>
  <div
    class="container"
  >
    <h2>
      Trips from Paris to Lyon
    </h2>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Num
          </th>
          <th
            scope="col"
          >
            From
          </th>
          <th
            scope="col"
          >
            To
          </th>
          <th
            scope="col"
          >
            Price
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            1
          </th>
          <td>
            Paris0
          </td>
          <td>
            Lyon0
          </td>
          <td>
            0 EUR
          </td>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            2
          </th>
          <td>
            Paris1
          </td>
          <td>
            Lyon1
          </td>
          <td>
            10 EUR
          </td>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            3
          </th>
          <td>
            Paris2
          </td>
          <td>
            Lyon2
          </td>
          <td>
            20 EUR
          </td>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            4
          </th>
          <td>
            Paris3
          </td>
          <td>
            Lyon3
          </td>
          <td>
            30 EUR
          </td>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            5
          </th>
          <td>
            Paris4
          </td>
          <td>
            Lyon4
          </td>
          <td>
            40 EUR
          </td>
        </tr>
      </tbody>
    </table>
    <ul
      class="pagination justify-content-center"
    >
      <li
        class="page-item disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          class="page-link"
          rel="prev"
          role="button"
          tabindex="0"
        >
          Previous
        </a>
      </li>
      <li
        class="page-item active"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          class="page-link"
          role="button"
          tabindex="0"
        >
          1
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          aria-label="Page 2"
          class="page-link"
          role="button"
          tabindex="0"
        >
          2
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          aria-label="Page 3"
          class="page-link"
          role="button"
          tabindex="0"
        >
          3
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          aria-label="Page 4"
          class="page-link"
          role="button"
          tabindex="0"
        >
          4
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          aria-label="Page 5"
          class="page-link"
          role="button"
          tabindex="0"
        >
          5
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          aria-label="Page 6"
          class="page-link"
          role="button"
          tabindex="0"
        >
          6
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          class="page-link"
          role="button"
          tabindex="0"
        >
          ...
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          aria-label="Page 20"
          class="page-link"
          role="button"
          tabindex="0"
        >
          20
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          aria-disabled="false"
          aria-label="Next page"
          class="page-link"
          rel="next"
          role="button"
          tabindex="0"
        >
          Next
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`renders without pagination 1`] = `
<div>
  <div
    class="container"
  >
    <h2>
      Trips from Paris to Lyon
    </h2>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Num
          </th>
          <th
            scope="col"
          >
            From
          </th>
          <th
            scope="col"
          >
            To
          </th>
          <th
            scope="col"
          >
            Price
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            1
          </th>
          <td>
            Paris
          </td>
          <td>
            Lyon
          </td>
          <td>
            10 EUR
          </td>
        </tr>
      </tbody>
    </table>
    <ul
      class="pagination justify-content-center"
    >
      <li
        class="page-item disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          class="page-link"
          rel="prev"
          role="button"
          tabindex="0"
        >
          Previous
        </a>
      </li>
      <li
        class="page-item active"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          class="page-link"
          role="button"
          tabindex="0"
        >
          1
        </a>
      </li>
      <li
        class="page-item disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Next page"
          class="page-link"
          rel="next"
          role="button"
          tabindex="0"
        >
          Next
        </a>
      </li>
    </ul>
  </div>
</div>
`;
